hcloud/action_waiter.go

```diff
@@ -0,0 +1,102 @@
+package hcloud
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/samber/lo"
+)
+
+type ActionWaiter interface {
+	WaitForActionsFunc(ctx context.Context, actions []*Action, handleUpdate func(update *Action)) ([]*Action, error)
+	WaitForActions(ctx context.Context, actions []*Action) ([]*Action, error)
+	WaitForAction(ctx context.Context, action *Action) (*Action, error)
+}
+
+var _ ActionWaiter = (*ActionClient)(nil)
+
+// WaitForActionsFunc implements ActionWaiter.
+func (c *ActionClient) WaitForActionsFunc(ctx context.Context, actions []*Action, handleUpdate func(update *Action)) ([]*Action, error) {
+	result := make([]*Action, 0, len(actions))
+
+	running := make(map[int64]struct{}, len(actions))
+	for _, action := range actions {
+		if action.Status == ActionStatusRunning {
+			running[action.ID] = struct{}{}
+		} else {
+			result = append(result, action)
+		}
+	}
```

Comment:
> Thinking about the Watch usecase, do you think we should emit handleUpdate calls for events which are not running at the start of the function?

Hmm, yeah, this might make sens if we think about tracking progress, but the `handleUpdate` name seem a bit confusing in that case, since the action is the same as the one passed by the user.

Reply:
Resolved